### PR TITLE
GSoC 2017 - Efficient Discretizations of PDE Operators

### PR DESCRIPTION
**Pull request by shivin9**
cc @ChrisRackauckas @dextorious

### CODE REVIEW

File: blog/_posts/2017-09-06-gsoc-derivative_operators.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+// we use CSS to left justify single line equations in code cells.
+displayAlign: 'center',
+"HTML-CSS": {
+styles: {'.MathJax_Display': {"margin": 0}},
+linebreaks: { automatic: true }
+}
+});
+</script>
+
+This project is an attempt towards building a PDE solver for JuliaDiffEq using the [Finite Difference Method](https://en.wikipedia.org/wiki/Finite_difference_method)(FDM) approach. We take up the FDM approach instead of [FEM](https://en.wikipedia.org/wiki/Finite_element_method) and [FVM](https://en.wikipedia.org/wiki/Finite_volume_method) as there are many toolboxes which already exist for FEM and FVM but not for FDM. Also, there are many use cases where the geometry of the problem is simple enough to be solved by FDM methods which are much faster due to their being able to avoid the bottleneck step of matrix multiplication by using Linear transformations to mimic the effect of a matrix multiplication.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-09-06-gsoc-derivative_operators.md]: "by using Linear transformations" I don't think that's correct. All matrix operations are linear transformations. The key here is avoiding large sparse matrices and using matrices with more structure.

File: blog/_posts/2017-09-06-gsoc-derivative_operators.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+But matrix multiplication is costly, therefore it would be preferable to have the linear operator of the double partial differential instead of the transformation matrix.
+It would look something like:-
+
+    function double_partial(x,dx)
+      for i in 2:length(dx)-1
+        dx[i] = -1*x[i-1] + 2*x[i] + -1*x[i+1]
+      end
+      dx[1] = 2*x[1] + -1*x[2]
+      dx[end] = -1*x[end-1] + 2*x[end]
+    end
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-09-06-gsoc-derivative_operators.md]: Might want to mention that this is O(n) operations as opposed to O(n^2), while still avoiding the overheads of sparse matrices.

File: blog/_posts/2017-09-06-gsoc-derivative_operators.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+So to convert the PDE into an ODE, we discritize the equation in space but not in time. Then this ODE can be solved efficiently by the existing solvers. Our heat equations transforms to the following ODE.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-09-06-gsoc-derivative_operators.md]: Will want to mention that this is the semilinear Heat equation

File: blog/_posts/2017-09-06-gsoc-derivative_operators.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+Now let us solve the solve the famous heat equation using the explicit discretization on a 2D `space x time` grid. The heat equation is:-
+        
+$$\frac{\partial u}{\partial t} - \frac{{\partial}^2 u}{\partial x^2} = 0$$
+
+For this example we consider a Dirichlet boundary condition with the initial distribution being parabolic. Since we have fixed the value at boundaries (in this case equal), after a long time we expect the 1D rod to be heated in a linear manner.
+
+        julia> using DiffEqOperators, DifferentialEquations, Plots
+        julia> x = collect(-pi : 2pi/511 : pi);
+        julia> u0 = -(x - 0.5).^2 + 1/12;
+        julia> A = DerivativeOperator{Float64}(2,2,2pi/511,512,:Dirichlet,:Dirichlet;BC=(u0[1],u0[end]));
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-09-06-gsoc-derivative_operators.md]: Explain this piece of code. Mention that this is for the second order approximation to the second derivative.

File: blog/_posts/2017-09-06-gsoc-derivative_operators.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+The general solution can then be written as follows:-
+$u_i^{n+1} = u_i^n - \Delta t[a^{+}u_x^{-} + a^{-}u_x^{+}]$
+
+The solution of the **KdV equation** using upwind operator looks better.
+
+![KdV solved using Upwind operators](/images/blog/2017-09-06-gsoc-derivative_operators/kdv_upwind.png)
+
+## Future Work
+Although vanilla `DerivativeOperators` and the `UpwindOperators` form the major part of DiffEqOperators there is still a lot to be done. A major functionality which is half implemented is application of DiffEqOperators on high dimensional spaces. Currently we support mixed and normal derivatives on 2D spaces only. There are open [issues](https://github.com/JuliaDiffEq/DiffEqOperators.jl/issues/20) and implementation [ideas](https://github.com/JuliaDiffEq/DiffEqOperators.jl/issues/21) on the issues page.
```
Comment (ChrisRackauckas) [inline blog/_posts/2017-09-06-gsoc-derivative_operators.md]: You should probably mention current issues: Robin BCs.

Comment (shivin9) [reply blog/_posts/2017-09-06-gsoc-derivative_operators.md]: yeah maybe that will be more proper.

**Comment by ChrisRackauckas** [comment]
I would also add a point on retreiving the stencil, `full(A)`, and `sparse(A)`. I think that the fact that this also gives you back the high order stencils will be a major selling point to many people who may need to do more complicated BCs but don't want to implement Fornberg to get there.

**Comment by shivin9** [comment]
maybe we can... but that seems more suited to be in the documentation

**Comment by ChrisRackauckas** [comment]
I think it's a big part of the advertisement that it's easy to get the stencils though. It just needs one line or so in here.

File: blog/_posts/2017-09-06-gsoc-derivative_operators.md (hunk 1 above)
Comment (shivin9) [reply blog/_posts/2017-09-06-gsoc-derivative_operators.md]: I meant to say that we are avoiding matrix multiplication by using a function to modify the vector which mimics the effect of matrix multiplication.

Comment (ChrisRackauckas) [reply blog/_posts/2017-09-06-gsoc-derivative_operators.md]: Yes, say it more directly.

File: blog/_posts/2017-09-06-gsoc-derivative_operators.md (hunk 5 above)
Comment (ChrisRackauckas) [reply blog/_posts/2017-09-06-gsoc-derivative_operators.md]: Cool. This done.

**Comment by ChrisRackauckas** [comment]
I see two comments left, and @dextorious should give a review (if available)

**Comment by ararslan** [comment]
screen shot 2017-09-10 at 9 35 39 am
Looks like something is wrong here. If anyone knows how to properly fix it, please submit a PR.

**Comment by ChrisRackauckas** [comment]
It's a missing bracket in the TeX.

**Comment by shivin9** [comment]
@ararslan fixed.

**Comment by ararslan** [comment]
Where did you fix it? You'll have to submit your changes as a PR here for it to take effect.

**Pull request merged**